### PR TITLE
Prevent concurrent access to shared LockUtil fields

### PR DESCRIPTION
**Pull request by ssoloff**
I came across this while debugging another [issue](https://github.com/triplea-game/triplea/pull/1477#issuecomment-273660108).  In summary, the `LockUtil` class uses a `static` field (`m_locksHeldWhenAcquired`) to track the order in which locks are acquired across all threads.  However, it uses an instance field to prevent concurrent access to `m_locksHeldWhenAcquired`.  There are multiple instances of `LockUtil` used in the application (e.g. `Tile.S_TILE_LOCKUTIL`, `GameData#lockUtil`, etc.).  Thus, different locks are being used by different `LockUtil` instances to synchronize, which may lead to concurrent access to `m_locksHeldWhenAcquired`.

In addition to the fix, I did a bit of refactoring around `LockUtil`.  Please advise if you'd like any of the refactorings backed out.

The highlights of this change are listed below:

* For all intents and purposes, the use of `static` fields in `LockUtil` effectively make it a singleton.  I went ahead and formalized this by fully implementing the singleton pattern in this class.
* I reduced the visibility of several `LockUtil` members (e.g. `ErrorReporter`), as they were only used for testing.
* Fields were renamed to conform to the project's [coding standard](http://www.triplea-game.org/dev_docs/dev/code_standards).
* Ensure access to the error reporter is atomic (it is unlikely this field will be accessed by multiple threads during testing, but just in case...); and ensure tests restore the original value of this field when complete.
* Remove unnecessary synchronization around `ThreadLocal` usage.
* Extract methods to acquire/release locks that hide the usage of `LockUtil` as much as possible (similar to what is done in `GameData`).

There are existing unit tests for `LockUtil`, which achieve nearly 100% coverage on the `LockUtil` methods.  However, this change seems somewhat risky given how much surface area `LockUtil` actually touches.  **I would appreciate any advice for additional tests (automated or manual) I could perform that would exercise a lot of concurrency around `GameData` and `Tile`s to ensure I didn't introduce any regressions.**

### CODE REVIEW

**Comment by ron-murhammer** [comment]
@ssoloff This needs rebased.

**Comment by ssoloff** [comment]
@ron-murhammer Rebased against 1d02356 (1.9.0.0.3712).

File: src/games/strategy/thread/LockUtil.java [hunk 1]
```diff
@@ -21,42 +24,54 @@
  * you are considering your ambitious multi-threaded code a mistake, and you are trying to limit the damage.
  * <p>
  */
-public class LockUtil {
+public final class LockUtil {
+  private static final LockUtil instance = new LockUtil();
```
Comment (DanVanAtta) [inline src/games/strategy/thread/LockUtil.java]: Prefer to use enum singleton's. We should have a number of examples, but basically make this class an enum with a single 'singleton' instance.

Comment (ssoloff) [inline src/games/strategy/thread/LockUtil.java]: You know, I must have read Bloch cover-to-cover at least three times, and I don't remember this gem (emphasis Bloch's):

> While this approach has yet to be widely adopted, **a single-element enum type is the best way to implement a singleton**.

Thanks for the nudge!

**Comment by DanVanAtta** [comment]
LGTM

**Pull request merged**